src/python/pants/backend/python/goals/setup_py.py

```diff
@@ -563,72 +560,46 @@ async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
-        entry_point = binary[PexEntryPointField].value
-        url = "https://python-packaging.readthedocs.io/en/latest/command-line-scripts.html#the-console-scripts-entry-point"
-        if not entry_point.function:
-            raise InvalidEntryPoint(
-                "Every `pex_binary` used in `with_binaries()` for the `provides()` field for "
-                f"{exported_addr} must end in the format `:my_func` for the `entry_point` field, "
-                f"but {binary.address} set it to {entry_point.spec!r}. For example, set "
-                f"`entry_point='{entry_point.module}:main'. See {url}."
-            )
-        entry_point_requests.append(ResolvePexEntryPointRequest(binary[PexEntryPointField]))
-
-    binary_entry_points = await MultiGet(
-        Get(ResolvedPexEntryPoint, ResolvePexEntryPointRequest, request)
-        for request in entry_point_requests
-    )
-
-    entry_points_from_with_binaries = {
-        key: binary_entry_point.val.spec
-        for key, binary_entry_point in zip(key_to_binary_spec.keys(), binary_entry_points)
-        if binary_entry_point.val is not None
-    }
-
-    # Collect entry points from `python_distribution(entry_points=...)`
-    entry_points_from_field = {}
-    if exported_target.target.has_field(PythonDistributionEntryPointsField):
-        entry_points_field = await Get(
+    # Resolve entry points from python_distribution(entry_points=...) and from
+    # python_distribution(provides=setup_py(entry_points=...).with_binaries(...)
+    entry_points_field, provides_field = await MultiGet(
+        Get(
             ResolvedPythonDistributionEntryPoints,
             ResolvePythonDistributionEntryPointsRequest(
-                exported_target.target[PythonDistributionEntryPointsField]
+                entry_points_field=exported_target.target.get(PythonDistributionEntryPointsField)
             ),
-        )
+        ),
+        Get(
+            ResolvedPythonDistributionEntryPoints,
+            ResolvePythonDistributionEntryPointsRequest(
+                provides_field=exported_target.target.get(PythonProvidesField)
+            ),
+        ),
+    )
 
-        entry_points_from_field = {
-            category: {ep_name: ep_val.entry_point.spec for ep_name, ep_val in entry_points.items()}
-            for category, entry_points in entry_points_field.val.items()
-        }
+    entry_points_from_field = {
+        category: {ep_name: ep_val.entry_point.spec for ep_name, ep_val in entry_points.items()}
+        for category, entry_points in entry_points_field.val.items()
+    }
 
-    # Collect any entry points from the setup_py() object. Note that this was already normalized in
-    # `python_artifacts.py`.
-    entry_points_from_provides = setup_kwargs.get("entry_points", {})
+    entry_points_from_provides = {
+        category: {ep_name: ep_val.entry_point.spec for ep_name, ep_val in entry_points.items()}
+        for category, entry_points in provides_field.val.items()
+    }
```

Comment:
Normally, we prefer comprehensions like this. But here, it looks like you're iterating over `entry_points.items()` more than necessary, and so [the style guide says](https://www.pantsbuild.org/docs/style-guide#prefer-comprehensions):

> If constructing multiple collections by iterating over the same original collection, use a for loop for performance.

Here, I think you would have `for ep_name, ep_val in entry_points.items()` in your outer loop. Then two inner loops over `provides_field.val.items()` and `entry_points_field.val.items()`.

Reply:
Eh.. it is not the same `entry_points.items()`, the first comes from `entry_points_field.val`, the other from `provides_field.val`. ;)

Could however have a little helper function to reduce code duplication, and this would become super clear :)